/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.opencb.biodata.models.variant.avro;  
@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
public class VariantAnnotation extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"VariantAnnotation\",\"namespace\":\"org.opencb.biodata.models.variant.avro\",\"fields\":[{\"name\":\"chromosome\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"start\",\"type\":[\"null\",\"int\"]},{\"name\":\"reference\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"alternate\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ancestralAllele\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"xrefs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Xref\",\"fields\":[{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"hgvs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"displayConsequenceType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"consequenceTypes\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ConsequenceType\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ensemblGeneId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ensemblTranscriptId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"strand\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"biotype\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"transcriptAnnotationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"cdnaPosition\",\"type\":[\"null\",\"int\"]},{\"name\":\"cdsPosition\",\"type\":[\"null\",\"int\"]},{\"name\":\"codon\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"proteinVariantAnnotation\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ProteinVariantAnnotation\",\"fields\":[{\"name\":\"uniprotAccession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"uniprotName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"position\",\"type\":\"int\"},{\"name\":\"reference\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"uniprotVariantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"functionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"substitutionScores\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Score\",\"fields\":[{\"name\":\"score\",\"type\":\"double\"},{\"name\":\"source\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"keywords\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"features\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ProteinFeature\",\"fields\":[{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"start\",\"type\":\"int\"},{\"name\":\"end\",\"type\":\"int\"},{\"name\":\"type\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]}]}]},{\"name\":\"sequenceOntologyTerms\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"SequenceOntologyTerm\",\"fields\":[{\"name\":\"accession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}}]}},\"default\":[]},{\"name\":\"populationFrequencies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PopulationFrequency\",\"fields\":[{\"name\":\"study\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"population\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"refAllele\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"altAllele\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"refAlleleFreq\",\"type\":\"float\"},{\"name\":\"altAlleleFreq\",\"type\":\"float\"},{\"name\":\"refHomGenotypeFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"hetGenotypeFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"altHomGenotypeFreq\",\"type\":[\"null\",\"float\"]}]}}]},{\"name\":\"minorAllele\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"minorAlleleFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"conservation\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Score\"}]},{\"name\":\"geneExpression\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Expression\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"transcriptId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"experimentalFactor\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"factorValue\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"experimentId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"technologyPlatform\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"expression\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ExpressionCall\",\"symbols\":[\"UP\",\"DOWN\"]}]},{\"name\":\"pvalue\",\"type\":[\"null\",\"float\"]}]}}]},{\"name\":\"geneTraitAssociation\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GeneTraitAssociation\",\"fields\":[{\"name\":\"id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"hpo\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"score\",\"type\":[\"null\",\"float\"]},{\"name\":\"numberOfPubmeds\",\"type\":[\"null\",\"int\"]},{\"name\":\"associationTypes\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"sources\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"source\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"geneDrugInteraction\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GeneDrugInteraction\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"drugName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"studyType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"type\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"variantTraitAssociation\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantTraitAssociation\",\"fields\":[{\"name\":\"clinvar\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ClinVar\",\"fields\":[{\"name\":\"accession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"clinicalSignificance\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"traits\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"geneNames\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"reviewStatus\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"gwas\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Gwas\",\"fields\":[{\"name\":\"snpIdCurrent\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"traits\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"riskAlleleFrequency\",\"type\":\"double\"},{\"name\":\"reportedGenes\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"cosmic\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Cosmic\",\"fields\":[{\"name\":\"mutationId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"primarySite\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"siteSubtype\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"primaryHistology\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"histologySubtype\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"sampleSource\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"tumourOrigin\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"","geneName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"mutationSomaticStatus\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]}]}]},{\"name\":\"functionalScore\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Score\"}]},{\"name\":\"additionalAttributes\",\"type\":[\"null\",{\"type\":\"map\",\"values\":[\"int\",\"float\",\"double\",{\"type\":\"string\",\"avro.java.string\":\"String\"},{\"type\":\"map\",\"values\":[\"int\",\"float\",\"double\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"avro.java.string\":\"String\"}],\"avro.java.string\":\"String\"}]}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
   private java.lang.String chromosome;
   private java.lang.Integer start;
   private java.lang.String reference;
   private java.lang.String alternate;
   private java.lang.String ancestralAllele;
   private java.lang.String id;
   private java.util.List<org.opencb.biodata.models.variant.avro.Xref> xrefs;
   private java.util.List<java.lang.String> hgvs;
   private java.lang.String displayConsequenceType;
   private java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> consequenceTypes;
   private java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> populationFrequencies;
   private java.lang.String minorAllele;
   private java.lang.Float minorAlleleFreq;
   private java.util.List<org.opencb.biodata.models.variant.avro.Score> conservation;
   private java.util.List<org.opencb.biodata.models.variant.avro.Expression> geneExpression;
   private java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> geneTraitAssociation;
   private java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> geneDrugInteraction;
   private org.opencb.biodata.models.variant.avro.VariantTraitAssociation variantTraitAssociation;
   private java.util.List<org.opencb.biodata.models.variant.avro.Score> functionalScore;
   private java.util.Map<java.lang.String,java.lang.Object> additionalAttributes;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>. 
   */
  public VariantAnnotation() {}

  /**
   * All-args constructor.
   */
  public VariantAnnotation(java.lang.String chromosome, java.lang.Integer start, java.lang.String reference, java.lang.String alternate, java.lang.String ancestralAllele, java.lang.String id, java.util.List<org.opencb.biodata.models.variant.avro.Xref> xrefs, java.util.List<java.lang.String> hgvs, java.lang.String displayConsequenceType, java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> consequenceTypes, java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> populationFrequencies, java.lang.String minorAllele, java.lang.Float minorAlleleFreq, java.util.List<org.opencb.biodata.models.variant.avro.Score> conservation, java.util.List<org.opencb.biodata.models.variant.avro.Expression> geneExpression, java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> geneTraitAssociation, java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> geneDrugInteraction, org.opencb.biodata.models.variant.avro.VariantTraitAssociation variantTraitAssociation, java.util.List<org.opencb.biodata.models.variant.avro.Score> functionalScore, java.util.Map<java.lang.String,java.lang.Object> additionalAttributes) {
    this.chromosome = chromosome;
    this.start = start;
    this.reference = reference;
    this.alternate = alternate;
    this.ancestralAllele = ancestralAllele;
    this.id = id;
    this.xrefs = xrefs;
    this.hgvs = hgvs;
    this.displayConsequenceType = displayConsequenceType;
    this.consequenceTypes = consequenceTypes;
    this.populationFrequencies = populationFrequencies;
    this.minorAllele = minorAllele;
    this.minorAlleleFreq = minorAlleleFreq;
    this.conservation = conservation;
    this.geneExpression = geneExpression;
    this.geneTraitAssociation = geneTraitAssociation;
    this.geneDrugInteraction = geneDrugInteraction;
    this.variantTraitAssociation = variantTraitAssociation;
    this.functionalScore = functionalScore;
    this.additionalAttributes = additionalAttributes;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call. 
  public java.lang.Object get(int field$) {
    switch (field$) {
    case 0: return chromosome;
    case 1: return start;
    case 2: return reference;
    case 3: return alternate;
    case 4: return ancestralAllele;
    case 5: return id;
    case 6: return xrefs;
    case 7: return hgvs;
    case 8: return displayConsequenceType;
    case 9: return consequenceTypes;
    case 10: return populationFrequencies;
    case 11: return minorAllele;
    case 12: return minorAlleleFreq;
    case 13: return conservation;
    case 14: return geneExpression;
    case 15: return geneTraitAssociation;
    case 16: return geneDrugInteraction;
    case 17: return variantTraitAssociation;
    case 18: return functionalScore;
    case 19: return additionalAttributes;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }
  // Used by DatumReader.  Applications should not call. 
  @SuppressWarnings(value="unchecked")
  public void put(int field$, java.lang.Object value$) {
    switch (field$) {
    case 0: chromosome = (java.lang.String)value$; break;
    case 1: start = (java.lang.Integer)value$; break;
    case 2: reference = (java.lang.String)value$; break;
    case 3: alternate = (java.lang.String)value$; break;
    case 4: ancestralAllele = (java.lang.String)value$; break;
    case 5: id = (java.lang.String)value$; break;
    case 6: xrefs = (java.util.List<org.opencb.biodata.models.variant.avro.Xref>)value$; break;
    case 7: hgvs = (java.util.List<java.lang.String>)value$; break;
    case 8: displayConsequenceType = (java.lang.String)value$; break;
    case 9: consequenceTypes = (java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType>)value$; break;
    case 10: populationFrequencies = (java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency>)value$; break;
    case 11: minorAllele = (java.lang.String)value$; break;
    case 12: minorAlleleFreq = (java.lang.Float)value$; break;
    case 13: conservation = (java.util.List<org.opencb.biodata.models.variant.avro.Score>)value$; break;
    case 14: geneExpression = (java.util.List<org.opencb.biodata.models.variant.avro.Expression>)value$; break;
    case 15: geneTraitAssociation = (java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation>)value$; break;
    case 16: geneDrugInteraction = (java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction>)value$; break;
    case 17: variantTraitAssociation = (org.opencb.biodata.models.variant.avro.VariantTraitAssociation)value$; break;
    case 18: functionalScore = (java.util.List<org.opencb.biodata.models.variant.avro.Score>)value$; break;
    case 19: additionalAttributes = (java.util.Map<java.lang.String,java.lang.Object>)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'chromosome' field.
   */
  public java.lang.String getChromosome() {
    return chromosome;
  }

  /**
   * Sets the value of the 'chromosome' field.
   * @param value the value to set.
   */
  public void setChromosome(java.lang.String value) {
    this.chromosome = value;
  }

  /**
   * Gets the value of the 'start' field.
   */
  public java.lang.Integer getStart() {
    return start;
  }

  /**
   * Sets the value of the 'start' field.
   * @param value the value to set.
   */
  public void setStart(java.lang.Integer value) {
    this.start = value;
  }

  /**
   * Gets the value of the 'reference' field.
   */
  public java.lang.String getReference() {
    return reference;
  }

  /**
   * Sets the value of the 'reference' field.
   * @param value the value to set.
   */
  public void setReference(java.lang.String value) {
    this.reference = value;
  }

  /**
   * Gets the value of the 'alternate' field.
   */
  public java.lang.String getAlternate() {
    return alternate;
  }

  /**
   * Sets the value of the 'alternate' field.
   * @param value the value to set.
   */
  public void setAlternate(java.lang.String value) {
    this.alternate = value;
  }

  /**
   * Gets the value of the 'ancestralAllele' field.
   */
  public java.lang.String getAncestralAllele() {
    return ancestralAllele;
  }

  /**
   * Sets the value of the 'ancestralAllele' field.
   * @param value the value to set.
   */
  public void setAncestralAllele(java.lang.String value) {
    this.ancestralAllele = value;
  }

  /**
   * Gets the value of the 'id' field.
   */
  public java.lang.String getId() {
    return id;
  }

  /**
   * Sets the value of the 'id' field.
   * @param value the value to set.
   */
  public void setId(java.lang.String value) {
    this.id = value;
  }

  /**
   * Gets the value of the 'xrefs' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.Xref> getXrefs() {
    return xrefs;
  }

  /**
   * Sets the value of the 'xrefs' field.
   * @param value the value to set.
   */
  public void setXrefs(java.util.List<org.opencb.biodata.models.variant.avro.Xref> value) {
    this.xrefs = value;
  }

  /**
   * Gets the value of the 'hgvs' field.
   */
  public java.util.List<java.lang.String> getHgvs() {
    return hgvs;
  }

  /**
   * Sets the value of the 'hgvs' field.
   * @param value the value to set.
   */
  public void setHgvs(java.util.List<java.lang.String> value) {
    this.hgvs = value;
  }

  /**
   * Gets the value of the 'displayConsequenceType' field.
   */
  public java.lang.String getDisplayConsequenceType() {
    return displayConsequenceType;
  }

  /**
   * Sets the value of the 'displayConsequenceType' field.
   * @param value the value to set.
   */
  public void setDisplayConsequenceType(java.lang.String value) {
    this.displayConsequenceType = value;
  }

  /**
   * Gets the value of the 'consequenceTypes' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> getConsequenceTypes() {
    return consequenceTypes;
  }

  /**
   * Sets the value of the 'consequenceTypes' field.
   * @param value the value to set.
   */
  public void setConsequenceTypes(java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> value) {
    this.consequenceTypes = value;
  }

  /**
   * Gets the value of the 'populationFrequencies' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> getPopulationFrequencies() {
    return populationFrequencies;
  }

  /**
   * Sets the value of the 'populationFrequencies' field.
   * @param value the value to set.
   */
  public void setPopulationFrequencies(java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> value) {
    this.populationFrequencies = value;
  }

  /**
   * Gets the value of the 'minorAllele' field.
   */
  public java.lang.String getMinorAllele() {
    return minorAllele;
  }

  /**
   * Sets the value of the 'minorAllele' field.
   * @param value the value to set.
   */
  public void setMinorAllele(java.lang.String value) {
    this.minorAllele = value;
  }

  /**
   * Gets the value of the 'minorAlleleFreq' field.
   */
  public java.lang.Float getMinorAlleleFreq() {
    return minorAlleleFreq;
  }

  /**
   * Sets the value of the 'minorAlleleFreq' field.
   * @param value the value to set.
   */
  public void setMinorAlleleFreq(java.lang.Float value) {
    this.minorAlleleFreq = value;
  }

  /**
   * Gets the value of the 'conservation' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.Score> getConservation() {
    return conservation;
  }

  /**
   * Sets the value of the 'conservation' field.
   * @param value the value to set.
   */
  public void setConservation(java.util.List<org.opencb.biodata.models.variant.avro.Score> value) {
    this.conservation = value;
  }

  /**
   * Gets the value of the 'geneExpression' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.Expression> getGeneExpression() {
    return geneExpression;
  }

  /**
   * Sets the value of the 'geneExpression' field.
   * @param value the value to set.
   */
  public void setGeneExpression(java.util.List<org.opencb.biodata.models.variant.avro.Expression> value) {
    this.geneExpression = value;
  }

  /**
   * Gets the value of the 'geneTraitAssociation' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> getGeneTraitAssociation() {
    return geneTraitAssociation;
  }

  /**
   * Sets the value of the 'geneTraitAssociation' field.
   * @param value the value to set.
   */
  public void setGeneTraitAssociation(java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> value) {
    this.geneTraitAssociation = value;
  }

  /**
   * Gets the value of the 'geneDrugInteraction' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> getGeneDrugInteraction() {
    return geneDrugInteraction;
  }

  /**
   * Sets the value of the 'geneDrugInteraction' field.
   * @param value the value to set.
   */
  public void setGeneDrugInteraction(java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> value) {
    this.geneDrugInteraction = value;
  }

  /**
   * Gets the value of the 'variantTraitAssociation' field.
   */
  public org.opencb.biodata.models.variant.avro.VariantTraitAssociation getVariantTraitAssociation() {
    return variantTraitAssociation;
  }

  /**
   * Sets the value of the 'variantTraitAssociation' field.
   * @param value the value to set.
   */
  public void setVariantTraitAssociation(org.opencb.biodata.models.variant.avro.VariantTraitAssociation value) {
    this.variantTraitAssociation = value;
  }

  /**
   * Gets the value of the 'functionalScore' field.
   */
  public java.util.List<org.opencb.biodata.models.variant.avro.Score> getFunctionalScore() {
    return functionalScore;
  }

  /**
   * Sets the value of the 'functionalScore' field.
   * @param value the value to set.
   */
  public void setFunctionalScore(java.util.List<org.opencb.biodata.models.variant.avro.Score> value) {
    this.functionalScore = value;
  }

  /**
   * Gets the value of the 'additionalAttributes' field.
   */
  public java.util.Map<java.lang.String,java.lang.Object> getAdditionalAttributes() {
    return additionalAttributes;
  }

  /**
   * Sets the value of the 'additionalAttributes' field.
   * @param value the value to set.
   */
  public void setAdditionalAttributes(java.util.Map<java.lang.String,java.lang.Object> value) {
    this.additionalAttributes = value;
  }

  /** Creates a new VariantAnnotation RecordBuilder */
  public static org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder newBuilder() {
    return new org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder();
  }
  
  /** Creates a new VariantAnnotation RecordBuilder by copying an existing Builder */
  public static org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder newBuilder(org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder other) {
    return new org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder(other);
  }
  
  /** Creates a new VariantAnnotation RecordBuilder by copying an existing VariantAnnotation instance */
  public static org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder newBuilder(org.opencb.biodata.models.variant.avro.VariantAnnotation other) {
    return new org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder(other);
  }
  
  /**
   * RecordBuilder for VariantAnnotation instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<VariantAnnotation>
    implements org.apache.avro.data.RecordBuilder<VariantAnnotation> {

    private java.lang.String chromosome;
    private java.lang.Integer start;
    private java.lang.String reference;
    private java.lang.String alternate;
    private java.lang.String ancestralAllele;
    private java.lang.String id;
    private java.util.List<org.opencb.biodata.models.variant.avro.Xref> xrefs;
    private java.util.List<java.lang.String> hgvs;
    private java.lang.String displayConsequenceType;
    private java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> consequenceTypes;
    private java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> populationFrequencies;
    private java.lang.String minorAllele;
    private java.lang.Float minorAlleleFreq;
    private java.util.List<org.opencb.biodata.models.variant.avro.Score> conservation;
    private java.util.List<org.opencb.biodata.models.variant.avro.Expression> geneExpression;
    private java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> geneTraitAssociation;
    private java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> geneDrugInteraction;
    private org.opencb.biodata.models.variant.avro.VariantTraitAssociation variantTraitAssociation;
    private java.util.List<org.opencb.biodata.models.variant.avro.Score> functionalScore;
    private java.util.Map<java.lang.String,java.lang.Object> additionalAttributes;

    /** Creates a new Builder */
    private Builder() {
      super(org.opencb.biodata.models.variant.avro.VariantAnnotation.SCHEMA$);
    }
    
    /** Creates a Builder by copying an existing Builder */
    private Builder(org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.chromosome)) {
        this.chromosome = data().deepCopy(fields()[0].schema(), other.chromosome);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.start)) {
        this.start = data().deepCopy(fields()[1].schema(), other.start);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.reference)) {
        this.reference = data().deepCopy(fields()[2].schema(), other.reference);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.alternate)) {
        this.alternate = data().deepCopy(fields()[3].schema(), other.alternate);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.ancestralAllele)) {
        this.ancestralAllele = data().deepCopy(fields()[4].schema(), other.ancestralAllele);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.id)) {
        this.id = data().deepCopy(fields()[5].schema(), other.id);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.xrefs)) {
        this.xrefs = data().deepCopy(fields()[6].schema(), other.xrefs);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.hgvs)) {
        this.hgvs = data().deepCopy(fields()[7].schema(), other.hgvs);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.displayConsequenceType)) {
        this.displayConsequenceType = data().deepCopy(fields()[8].schema(), other.displayConsequenceType);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.consequenceTypes)) {
        this.consequenceTypes = data().deepCopy(fields()[9].schema(), other.consequenceTypes);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.populationFrequencies)) {
        this.populationFrequencies = data().deepCopy(fields()[10].schema(), other.populationFrequencies);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.minorAllele)) {
        this.minorAllele = data().deepCopy(fields()[11].schema(), other.minorAllele);
        fieldSetFlags()[11] = true;
      }
      if (isValidValue(fields()[12], other.minorAlleleFreq)) {
        this.minorAlleleFreq = data().deepCopy(fields()[12].schema(), other.minorAlleleFreq);
        fieldSetFlags()[12] = true;
      }
      if (isValidValue(fields()[13], other.conservation)) {
        this.conservation = data().deepCopy(fields()[13].schema(), other.conservation);
        fieldSetFlags()[13] = true;
      }
      if (isValidValue(fields()[14], other.geneExpression)) {
        this.geneExpression = data().deepCopy(fields()[14].schema(), other.geneExpression);
        fieldSetFlags()[14] = true;
      }
      if (isValidValue(fields()[15], other.geneTraitAssociation)) {
        this.geneTraitAssociation = data().deepCopy(fields()[15].schema(), other.geneTraitAssociation);
        fieldSetFlags()[15] = true;
      }
      if (isValidValue(fields()[16], other.geneDrugInteraction)) {
        this.geneDrugInteraction = data().deepCopy(fields()[16].schema(), other.geneDrugInteraction);
        fieldSetFlags()[16] = true;
      }
      if (isValidValue(fields()[17], other.variantTraitAssociation)) {
        this.variantTraitAssociation = data().deepCopy(fields()[17].schema(), other.variantTraitAssociation);
        fieldSetFlags()[17] = true;
      }
      if (isValidValue(fields()[18], other.functionalScore)) {
        this.functionalScore = data().deepCopy(fields()[18].schema(), other.functionalScore);
        fieldSetFlags()[18] = true;
      }
      if (isValidValue(fields()[19], other.additionalAttributes)) {
        this.additionalAttributes = data().deepCopy(fields()[19].schema(), other.additionalAttributes);
        fieldSetFlags()[19] = true;
      }
    }
    
    /** Creates a Builder by copying an existing VariantAnnotation instance */
    private Builder(org.opencb.biodata.models.variant.avro.VariantAnnotation other) {
            super(org.opencb.biodata.models.variant.avro.VariantAnnotation.SCHEMA$);
      if (isValidValue(fields()[0], other.chromosome)) {
        this.chromosome = data().deepCopy(fields()[0].schema(), other.chromosome);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.start)) {
        this.start = data().deepCopy(fields()[1].schema(), other.start);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.reference)) {
        this.reference = data().deepCopy(fields()[2].schema(), other.reference);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.alternate)) {
        this.alternate = data().deepCopy(fields()[3].schema(), other.alternate);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.ancestralAllele)) {
        this.ancestralAllele = data().deepCopy(fields()[4].schema(), other.ancestralAllele);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.id)) {
        this.id = data().deepCopy(fields()[5].schema(), other.id);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.xrefs)) {
        this.xrefs = data().deepCopy(fields()[6].schema(), other.xrefs);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.hgvs)) {
        this.hgvs = data().deepCopy(fields()[7].schema(), other.hgvs);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.displayConsequenceType)) {
        this.displayConsequenceType = data().deepCopy(fields()[8].schema(), other.displayConsequenceType);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.consequenceTypes)) {
        this.consequenceTypes = data().deepCopy(fields()[9].schema(), other.consequenceTypes);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.populationFrequencies)) {
        this.populationFrequencies = data().deepCopy(fields()[10].schema(), other.populationFrequencies);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.minorAllele)) {
        this.minorAllele = data().deepCopy(fields()[11].schema(), other.minorAllele);
        fieldSetFlags()[11] = true;
      }
      if (isValidValue(fields()[12], other.minorAlleleFreq)) {
        this.minorAlleleFreq = data().deepCopy(fields()[12].schema(), other.minorAlleleFreq);
        fieldSetFlags()[12] = true;
      }
      if (isValidValue(fields()[13], other.conservation)) {
        this.conservation = data().deepCopy(fields()[13].schema(), other.conservation);
        fieldSetFlags()[13] = true;
      }
      if (isValidValue(fields()[14], other.geneExpression)) {
        this.geneExpression = data().deepCopy(fields()[14].schema(), other.geneExpression);
        fieldSetFlags()[14] = true;
      }
      if (isValidValue(fields()[15], other.geneTraitAssociation)) {
        this.geneTraitAssociation = data().deepCopy(fields()[15].schema(), other.geneTraitAssociation);
        fieldSetFlags()[15] = true;
      }
      if (isValidValue(fields()[16], other.geneDrugInteraction)) {
        this.geneDrugInteraction = data().deepCopy(fields()[16].schema(), other.geneDrugInteraction);
        fieldSetFlags()[16] = true;
      }
      if (isValidValue(fields()[17], other.variantTraitAssociation)) {
        this.variantTraitAssociation = data().deepCopy(fields()[17].schema(), other.variantTraitAssociation);
        fieldSetFlags()[17] = true;
      }
      if (isValidValue(fields()[18], other.functionalScore)) {
        this.functionalScore = data().deepCopy(fields()[18].schema(), other.functionalScore);
        fieldSetFlags()[18] = true;
      }
      if (isValidValue(fields()[19], other.additionalAttributes)) {
        this.additionalAttributes = data().deepCopy(fields()[19].schema(), other.additionalAttributes);
        fieldSetFlags()[19] = true;
      }
    }

    /** Gets the value of the 'chromosome' field */
    public java.lang.String getChromosome() {
      return chromosome;
    }
    
    /** Sets the value of the 'chromosome' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setChromosome(java.lang.String value) {
      validate(fields()[0], value);
      this.chromosome = value;
      fieldSetFlags()[0] = true;
      return this; 
    }
    
    /** Checks whether the 'chromosome' field has been set */
    public boolean hasChromosome() {
      return fieldSetFlags()[0];
    }
    
    /** Clears the value of the 'chromosome' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearChromosome() {
      chromosome = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    /** Gets the value of the 'start' field */
    public java.lang.Integer getStart() {
      return start;
    }
    
    /** Sets the value of the 'start' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setStart(java.lang.Integer value) {
      validate(fields()[1], value);
      this.start = value;
      fieldSetFlags()[1] = true;
      return this; 
    }
    
    /** Checks whether the 'start' field has been set */
    public boolean hasStart() {
      return fieldSetFlags()[1];
    }
    
    /** Clears the value of the 'start' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearStart() {
      start = null;
      fieldSetFlags()[1] = false;
      return this;
    }

    /** Gets the value of the 'reference' field */
    public java.lang.String getReference() {
      return reference;
    }
    
    /** Sets the value of the 'reference' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setReference(java.lang.String value) {
      validate(fields()[2], value);
      this.reference = value;
      fieldSetFlags()[2] = true;
      return this; 
    }
    
    /** Checks whether the 'reference' field has been set */
    public boolean hasReference() {
      return fieldSetFlags()[2];
    }
    
    /** Clears the value of the 'reference' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearReference() {
      reference = null;
      fieldSetFlags()[2] = false;
      return this;
    }

    /** Gets the value of the 'alternate' field */
    public java.lang.String getAlternate() {
      return alternate;
    }
    
    /** Sets the value of the 'alternate' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setAlternate(java.lang.String value) {
      validate(fields()[3], value);
      this.alternate = value;
      fieldSetFlags()[3] = true;
      return this; 
    }
    
    /** Checks whether the 'alternate' field has been set */
    public boolean hasAlternate() {
      return fieldSetFlags()[3];
    }
    
    /** Clears the value of the 'alternate' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearAlternate() {
      alternate = null;
      fieldSetFlags()[3] = false;
      return this;
    }

    /** Gets the value of the 'ancestralAllele' field */
    public java.lang.String getAncestralAllele() {
      return ancestralAllele;
    }
    
    /** Sets the value of the 'ancestralAllele' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setAncestralAllele(java.lang.String value) {
      validate(fields()[4], value);
      this.ancestralAllele = value;
      fieldSetFlags()[4] = true;
      return this; 
    }
    
    /** Checks whether the 'ancestralAllele' field has been set */
    public boolean hasAncestralAllele() {
      return fieldSetFlags()[4];
    }
    
    /** Clears the value of the 'ancestralAllele' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearAncestralAllele() {
      ancestralAllele = null;
      fieldSetFlags()[4] = false;
      return this;
    }

    /** Gets the value of the 'id' field */
    public java.lang.String getId() {
      return id;
    }
    
    /** Sets the value of the 'id' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setId(java.lang.String value) {
      validate(fields()[5], value);
      this.id = value;
      fieldSetFlags()[5] = true;
      return this; 
    }
    
    /** Checks whether the 'id' field has been set */
    public boolean hasId() {
      return fieldSetFlags()[5];
    }
    
    /** Clears the value of the 'id' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearId() {
      id = null;
      fieldSetFlags()[5] = false;
      return this;
    }

    /** Gets the value of the 'xrefs' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.Xref> getXrefs() {
      return xrefs;
    }
    
    /** Sets the value of the 'xrefs' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setXrefs(java.util.List<org.opencb.biodata.models.variant.avro.Xref> value) {
      validate(fields()[6], value);
      this.xrefs = value;
      fieldSetFlags()[6] = true;
      return this; 
    }
    
    /** Checks whether the 'xrefs' field has been set */
    public boolean hasXrefs() {
      return fieldSetFlags()[6];
    }
    
    /** Clears the value of the 'xrefs' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearXrefs() {
      xrefs = null;
      fieldSetFlags()[6] = false;
      return this;
    }

    /** Gets the value of the 'hgvs' field */
    public java.util.List<java.lang.String> getHgvs() {
      return hgvs;
    }
    
    /** Sets the value of the 'hgvs' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setHgvs(java.util.List<java.lang.String> value) {
      validate(fields()[7], value);
      this.hgvs = value;
      fieldSetFlags()[7] = true;
      return this; 
    }
    
    /** Checks whether the 'hgvs' field has been set */
    public boolean hasHgvs() {
      return fieldSetFlags()[7];
    }
    
    /** Clears the value of the 'hgvs' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearHgvs() {
      hgvs = null;
      fieldSetFlags()[7] = false;
      return this;
    }

    /** Gets the value of the 'displayConsequenceType' field */
    public java.lang.String getDisplayConsequenceType() {
      return displayConsequenceType;
    }
    
    /** Sets the value of the 'displayConsequenceType' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setDisplayConsequenceType(java.lang.String value) {
      validate(fields()[8], value);
      this.displayConsequenceType = value;
      fieldSetFlags()[8] = true;
      return this; 
    }
    
    /** Checks whether the 'displayConsequenceType' field has been set */
    public boolean hasDisplayConsequenceType() {
      return fieldSetFlags()[8];
    }
    
    /** Clears the value of the 'displayConsequenceType' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearDisplayConsequenceType() {
      displayConsequenceType = null;
      fieldSetFlags()[8] = false;
      return this;
    }

    /** Gets the value of the 'consequenceTypes' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> getConsequenceTypes() {
      return consequenceTypes;
    }
    
    /** Sets the value of the 'consequenceTypes' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setConsequenceTypes(java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType> value) {
      validate(fields()[9], value);
      this.consequenceTypes = value;
      fieldSetFlags()[9] = true;
      return this; 
    }
    
    /** Checks whether the 'consequenceTypes' field has been set */
    public boolean hasConsequenceTypes() {
      return fieldSetFlags()[9];
    }
    
    /** Clears the value of the 'consequenceTypes' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearConsequenceTypes() {
      consequenceTypes = null;
      fieldSetFlags()[9] = false;
      return this;
    }

    /** Gets the value of the 'populationFrequencies' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> getPopulationFrequencies() {
      return populationFrequencies;
    }
    
    /** Sets the value of the 'populationFrequencies' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setPopulationFrequencies(java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency> value) {
      validate(fields()[10], value);
      this.populationFrequencies = value;
      fieldSetFlags()[10] = true;
      return this; 
    }
    
    /** Checks whether the 'populationFrequencies' field has been set */
    public boolean hasPopulationFrequencies() {
      return fieldSetFlags()[10];
    }
    
    /** Clears the value of the 'populationFrequencies' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearPopulationFrequencies() {
      populationFrequencies = null;
      fieldSetFlags()[10] = false;
      return this;
    }

    /** Gets the value of the 'minorAllele' field */
    public java.lang.String getMinorAllele() {
      return minorAllele;
    }
    
    /** Sets the value of the 'minorAllele' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setMinorAllele(java.lang.String value) {
      validate(fields()[11], value);
      this.minorAllele = value;
      fieldSetFlags()[11] = true;
      return this; 
    }
    
    /** Checks whether the 'minorAllele' field has been set */
    public boolean hasMinorAllele() {
      return fieldSetFlags()[11];
    }
    
    /** Clears the value of the 'minorAllele' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearMinorAllele() {
      minorAllele = null;
      fieldSetFlags()[11] = false;
      return this;
    }

    /** Gets the value of the 'minorAlleleFreq' field */
    public java.lang.Float getMinorAlleleFreq() {
      return minorAlleleFreq;
    }
    
    /** Sets the value of the 'minorAlleleFreq' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setMinorAlleleFreq(java.lang.Float value) {
      validate(fields()[12], value);
      this.minorAlleleFreq = value;
      fieldSetFlags()[12] = true;
      return this; 
    }
    
    /** Checks whether the 'minorAlleleFreq' field has been set */
    public boolean hasMinorAlleleFreq() {
      return fieldSetFlags()[12];
    }
    
    /** Clears the value of the 'minorAlleleFreq' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearMinorAlleleFreq() {
      minorAlleleFreq = null;
      fieldSetFlags()[12] = false;
      return this;
    }

    /** Gets the value of the 'conservation' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.Score> getConservation() {
      return conservation;
    }
    
    /** Sets the value of the 'conservation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setConservation(java.util.List<org.opencb.biodata.models.variant.avro.Score> value) {
      validate(fields()[13], value);
      this.conservation = value;
      fieldSetFlags()[13] = true;
      return this; 
    }
    
    /** Checks whether the 'conservation' field has been set */
    public boolean hasConservation() {
      return fieldSetFlags()[13];
    }
    
    /** Clears the value of the 'conservation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearConservation() {
      conservation = null;
      fieldSetFlags()[13] = false;
      return this;
    }

    /** Gets the value of the 'geneExpression' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.Expression> getGeneExpression() {
      return geneExpression;
    }
    
    /** Sets the value of the 'geneExpression' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setGeneExpression(java.util.List<org.opencb.biodata.models.variant.avro.Expression> value) {
      validate(fields()[14], value);
      this.geneExpression = value;
      fieldSetFlags()[14] = true;
      return this; 
    }
    
    /** Checks whether the 'geneExpression' field has been set */
    public boolean hasGeneExpression() {
      return fieldSetFlags()[14];
    }
    
    /** Clears the value of the 'geneExpression' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearGeneExpression() {
      geneExpression = null;
      fieldSetFlags()[14] = false;
      return this;
    }

    /** Gets the value of the 'geneTraitAssociation' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> getGeneTraitAssociation() {
      return geneTraitAssociation;
    }
    
    /** Sets the value of the 'geneTraitAssociation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setGeneTraitAssociation(java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation> value) {
      validate(fields()[15], value);
      this.geneTraitAssociation = value;
      fieldSetFlags()[15] = true;
      return this; 
    }
    
    /** Checks whether the 'geneTraitAssociation' field has been set */
    public boolean hasGeneTraitAssociation() {
      return fieldSetFlags()[15];
    }
    
    /** Clears the value of the 'geneTraitAssociation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearGeneTraitAssociation() {
      geneTraitAssociation = null;
      fieldSetFlags()[15] = false;
      return this;
    }

    /** Gets the value of the 'geneDrugInteraction' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> getGeneDrugInteraction() {
      return geneDrugInteraction;
    }
    
    /** Sets the value of the 'geneDrugInteraction' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setGeneDrugInteraction(java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction> value) {
      validate(fields()[16], value);
      this.geneDrugInteraction = value;
      fieldSetFlags()[16] = true;
      return this; 
    }
    
    /** Checks whether the 'geneDrugInteraction' field has been set */
    public boolean hasGeneDrugInteraction() {
      return fieldSetFlags()[16];
    }
    
    /** Clears the value of the 'geneDrugInteraction' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearGeneDrugInteraction() {
      geneDrugInteraction = null;
      fieldSetFlags()[16] = false;
      return this;
    }

    /** Gets the value of the 'variantTraitAssociation' field */
    public org.opencb.biodata.models.variant.avro.VariantTraitAssociation getVariantTraitAssociation() {
      return variantTraitAssociation;
    }
    
    /** Sets the value of the 'variantTraitAssociation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setVariantTraitAssociation(org.opencb.biodata.models.variant.avro.VariantTraitAssociation value) {
      validate(fields()[17], value);
      this.variantTraitAssociation = value;
      fieldSetFlags()[17] = true;
      return this; 
    }
    
    /** Checks whether the 'variantTraitAssociation' field has been set */
    public boolean hasVariantTraitAssociation() {
      return fieldSetFlags()[17];
    }
    
    /** Clears the value of the 'variantTraitAssociation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearVariantTraitAssociation() {
      variantTraitAssociation = null;
      fieldSetFlags()[17] = false;
      return this;
    }

    /** Gets the value of the 'functionalScore' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.Score> getFunctionalScore() {
      return functionalScore;
    }
    
    /** Sets the value of the 'functionalScore' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setFunctionalScore(java.util.List<org.opencb.biodata.models.variant.avro.Score> value) {
      validate(fields()[18], value);
      this.functionalScore = value;
      fieldSetFlags()[18] = true;
      return this; 
    }
    
    /** Checks whether the 'functionalScore' field has been set */
    public boolean hasFunctionalScore() {
      return fieldSetFlags()[18];
    }
    
    /** Clears the value of the 'functionalScore' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearFunctionalScore() {
      functionalScore = null;
      fieldSetFlags()[18] = false;
      return this;
    }

    /** Gets the value of the 'additionalAttributes' field */
    public java.util.Map<java.lang.String,java.lang.Object> getAdditionalAttributes() {
      return additionalAttributes;
    }
    
    /** Sets the value of the 'additionalAttributes' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder setAdditionalAttributes(java.util.Map<java.lang.String,java.lang.Object> value) {
      validate(fields()[19], value);
      this.additionalAttributes = value;
      fieldSetFlags()[19] = true;
      return this; 
    }
    
    /** Checks whether the 'additionalAttributes' field has been set */
    public boolean hasAdditionalAttributes() {
      return fieldSetFlags()[19];
    }
    
    /** Clears the value of the 'additionalAttributes' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation.Builder clearAdditionalAttributes() {
      additionalAttributes = null;
      fieldSetFlags()[19] = false;
      return this;
    }

    @Override
    public VariantAnnotation build() {
      try {
        VariantAnnotation record = new VariantAnnotation();
        record.chromosome = fieldSetFlags()[0] ? this.chromosome : (java.lang.String) defaultValue(fields()[0]);
        record.start = fieldSetFlags()[1] ? this.start : (java.lang.Integer) defaultValue(fields()[1]);
        record.reference = fieldSetFlags()[2] ? this.reference : (java.lang.String) defaultValue(fields()[2]);
        record.alternate = fieldSetFlags()[3] ? this.alternate : (java.lang.String) defaultValue(fields()[3]);
        record.ancestralAllele = fieldSetFlags()[4] ? this.ancestralAllele : (java.lang.String) defaultValue(fields()[4]);
        record.id = fieldSetFlags()[5] ? this.id : (java.lang.String) defaultValue(fields()[5]);
        record.xrefs = fieldSetFlags()[6] ? this.xrefs : (java.util.List<org.opencb.biodata.models.variant.avro.Xref>) defaultValue(fields()[6]);
        record.hgvs = fieldSetFlags()[7] ? this.hgvs : (java.util.List<java.lang.String>) defaultValue(fields()[7]);
        record.displayConsequenceType = fieldSetFlags()[8] ? this.displayConsequenceType : (java.lang.String) defaultValue(fields()[8]);
        record.consequenceTypes = fieldSetFlags()[9] ? this.consequenceTypes : (java.util.List<org.opencb.biodata.models.variant.avro.ConsequenceType>) defaultValue(fields()[9]);
        record.populationFrequencies = fieldSetFlags()[10] ? this.populationFrequencies : (java.util.List<org.opencb.biodata.models.variant.avro.PopulationFrequency>) defaultValue(fields()[10]);
        record.minorAllele = fieldSetFlags()[11] ? this.minorAllele : (java.lang.String) defaultValue(fields()[11]);
        record.minorAlleleFreq = fieldSetFlags()[12] ? this.minorAlleleFreq : (java.lang.Float) defaultValue(fields()[12]);
        record.conservation = fieldSetFlags()[13] ? this.conservation : (java.util.List<org.opencb.biodata.models.variant.avro.Score>) defaultValue(fields()[13]);
        record.geneExpression = fieldSetFlags()[14] ? this.geneExpression : (java.util.List<org.opencb.biodata.models.variant.avro.Expression>) defaultValue(fields()[14]);
        record.geneTraitAssociation = fieldSetFlags()[15] ? this.geneTraitAssociation : (java.util.List<org.opencb.biodata.models.variant.avro.GeneTraitAssociation>) defaultValue(fields()[15]);
        record.geneDrugInteraction = fieldSetFlags()[16] ? this.geneDrugInteraction : (java.util.List<org.opencb.biodata.models.variant.avro.GeneDrugInteraction>) defaultValue(fields()[16]);
        record.variantTraitAssociation = fieldSetFlags()[17] ? this.variantTraitAssociation : (org.opencb.biodata.models.variant.avro.VariantTraitAssociation) defaultValue(fields()[17]);
        record.functionalScore = fieldSetFlags()[18] ? this.functionalScore : (java.util.List<org.opencb.biodata.models.variant.avro.Score>) defaultValue(fields()[18]);
        record.additionalAttributes = fieldSetFlags()[19] ? this.additionalAttributes : (java.util.Map<java.lang.String,java.lang.Object>) defaultValue(fields()[19]);
        return record;
      } catch (Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }
}
